### Importing Library:

In [1]:
import pandas as pd

### Loading "Matches" Dataset:

In [2]:
data1 = pd.read_csv("matches.csv")
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
data1.head()

id  season       city        date                        team1  \
0   1    2017  Hyderabad  2017-04-05          Sunrisers Hyderabad   
1   2    2017       Pune  2017-04-06               Mumbai Indians   
2   3    2017     Rajkot  2017-04-07                Gujarat Lions   
3   4    2017     Indore  2017-04-08       Rising Pune Supergiant   
4   5    2017  Bangalore  2017-04-08  Royal Challengers Bangalore   

                         team2                  toss_winner toss_decision  \
0  Royal Challengers Bangalore  Royal Challengers Bangalore         field   
1       Rising Pune Supergiant       Rising Pune Supergiant         field   
2        Kolkata Knight Riders        Kolkata Knight Riders         field   
3              Kings XI Punjab              Kings XI Punjab         field   
4             Delhi Daredevils  Royal Challengers Bangalore           bat   

   result  dl_applied                       winner  win_by_runs  \
0  normal           0          Sunrisers Hyderabad           35   
1  normal           0       Rising Pune Supergiant            0   
2  normal           0        Kolkata Knight Riders            0   
3  normal           0              Kings XI Punjab            0   
4  normal           0  Royal Challengers Bangalore           15   

   win_by_wickets player_of_match                                      venue  \
0               0    Yuvraj Singh  Rajiv Gandhi International Stadium, Uppal   
1               7       SPD Smith    Maharashtra Cricket Association Stadium   
2              10         CA Lynn     Saurashtra Cricket Association Stadium   
3               6      GJ Maxwell                     Holkar Cricket Stadium   
4               0       KM Jadhav                      M Chinnaswamy Stadium   

          umpire1        umpire2 umpire3  
0     AY Dandekar       NJ Llong     NaN  
1  A Nand Kishore         S Ravi     NaN  
2     Nitin Menon      CK Nandan     NaN  
3    AK Chaudhary  C Shamshuddin     NaN  
4             NaN            NaN     NaN

### Replacing `Delhi Daredevils` with `Delhi Capitals`:

In [3]:
data1 = data1.replace("Delhi Daredevils", "Delhi Capitals")

### Filtering Delhi and CSK matches only:

In [4]:
csk_dc = data1.loc[((data1['team1'] == 'Chennai Super Kings') & (data1['team2'] == 'Delhi Capitals') | (data1['team1'] == 'Delhi Capitals') & (data1['team2'] == 'Chennai Super Kings'))]
csk_dc.head(10)

id  season          city        date                team1  \
78    79    2008       Chennai  2008-05-02  Chennai Super Kings   
86    87    2008         Delhi  2008-05-08       Delhi Capitals   
124  125    2009        Durban  2009-04-23       Delhi Capitals   
140  141    2009  Johannesburg  2009-05-02  Chennai Super Kings   
184  185    2010         Delhi  2010-03-19       Delhi Capitals   
223  224    2010       Chennai  2010-04-15  Chennai Super Kings   
288  289    2011       Chennai  2011-05-12  Chennai Super Kings   
317  318    2012         Delhi  2012-04-10  Chennai Super Kings   
363  364    2012       Chennai  2012-05-12       Delhi Capitals   
379  380    2012       Chennai  2012-05-25  Chennai Super Kings   

                   team2          toss_winner toss_decision  result  \
78        Delhi Capitals  Chennai Super Kings           bat  normal   
86   Chennai Super Kings  Chennai Super Kings         field  normal   
124  Chennai Super Kings       Delhi Capitals           bat  normal   
140       Delhi Capitals       Delhi Capitals         field  normal   
184  Chennai Super Kings       Delhi Capitals           bat  normal   
223       Delhi Capitals  Chennai Super Kings           bat  normal   
288       Delhi Capitals  Chennai Super Kings           bat  normal   
317       Delhi Capitals       Delhi Capitals         field  normal   
363  Chennai Super Kings  Chennai Super Kings         field  normal   
379       Delhi Capitals       Delhi Capitals         field  normal   

     dl_applied               winner  win_by_runs  win_by_wickets  \
78            0       Delhi Capitals            0               8   
86            0  Chennai Super Kings            0               4   
124           0       Delhi Capitals            9               0   
140           0  Chennai Super Kings           18               0   
184           0  Chennai Super Kings            0               5   
223           0       Delhi Capitals            0               6   
288           0  Chennai Super Kings           18               0   
317           0       Delhi Capitals            0               8   
363           0  Chennai Super Kings            0               9   
379           0  Chennai Super Kings           86               0   

    player_of_match                            venue          umpire1  \
78         V Sehwag  MA Chidambaram Stadium, Chepauk        BF Bowden   
86         MS Dhoni                 Feroz Shah Kotla        Aleem Dar   
124  AB de Villiers                        Kingsmead      BR Doctrove   
140       SB Jakati            New Wanderers Stadium        DJ Harper   
184       ML Hayden                 Feroz Shah Kotla      BR Doctrove   
223       G Gambhir  MA Chidambaram Stadium, Chepauk  HDPK Dharmasena   
288        MS Dhoni  MA Chidambaram Stadium, Chepauk        AM Saheba   
317        M Morkel                 Feroz Shah Kotla        Asad Rauf   
363   BW Hilfenhaus  MA Chidambaram Stadium, Chepauk            S Das   
379         M Vijay  MA Chidambaram Stadium, Chepauk      BR Doctrove   

         umpire2 umpire3  
78   K Hariharan     NaN  
86     RB Tiffin     NaN  
124   SJA Taufel     NaN  
140  RE Koertzen     NaN  
184  SK Tarapore     NaN  
223    SS Hazare     NaN  
288   SL Shastri     NaN  
317  SK Tarapore     NaN  
363  BR Doctrove     NaN  
379   SJA Taufel     NaN

## Question 1: Who will win the match?

#### Solution: Checking counts of winner:

In [7]:
csk_dc['winner'].value_counts()/csk_dc.shape[0] * 100

Chennai Super Kings    71.428571
Delhi Capitals         28.571429
Name: winner, dtype: float64

**Answer:** As per the above analysis, winning percentage of CSK is **71.43** over DC. But, in this season, DC is performing well and their team combination is also nice. I would predict that DC will win today's match.

## Question 2: How many runs will CSK score in the match?

#### Solution: Observing runs of CSK in each match:

### Loading "Deliveries" dataset:

In [8]:
data2 = pd.read_csv("deliveries.csv")
data2.head()

match_id  inning         batting_team                 bowling_team  over  \
0         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
1         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
2         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
3         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   
4         1       1  Sunrisers Hyderabad  Royal Challengers Bangalore     1   

   ball    batsman non_striker    bowler  is_super_over  wide_runs  bye_runs  \
0     1  DA Warner    S Dhawan  TS Mills              0          0         0   
1     2  DA Warner    S Dhawan  TS Mills              0          0         0   
2     3  DA Warner    S Dhawan  TS Mills              0          0         0   
3     4  DA Warner    S Dhawan  TS Mills              0          0         0   
4     5  DA Warner    S Dhawan  TS Mills              0          2         0   

   legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
0            0            0             0             0           0   
1            0            0             0             0           0   
2            0            0             0             4           0   
3            0            0             0             0           0   
4            0            0             0             0           2   

   total_runs player_dismissed dismissal_kind fielder  
0           0              NaN            NaN     NaN  
1           0              NaN            NaN     NaN  
2           4              NaN            NaN     NaN  
3           0              NaN            NaN     NaN  
4           2              NaN            NaN     NaN

### Replacing `Delhi Daredevils` with `Delhi Capitals`:

In [9]:
data2 = data2.replace("Delhi Daredevils", "Delhi Capitals")

### Filtering CSK and Delhi matches where batting team is CSK and bowling team is DC:

In [11]:
csk_runs = data2.loc[(data2['batting_team'] == "Chennai Super Kings") & (data2['bowling_team'] == "Delhi Capitals")]
csk_runs = pd.DataFrame(csk_runs.groupby('match_id')['total_runs'].sum())
csk_runs

total_runs
match_id            
79               169
87               188
125              180
141              163
185              190
224              112
289              176
318              110
364              115
380              222
404              169
446              168
465              177
483              181
519              150
565              119
7923             218
7945             130
11141            157
11341            186
11414            162

### Average CSK runs:

In [12]:
avg_csk_runs = csk_runs['total_runs'].mean()
avg_csk_runs

163.9047619047619

**Answer:** As per the above analysis, average score of CSK is **~ 164**.

## Question 3: How many wickets will CSK take from the 91st ball to the 120th ball?

#### Solution: Observing match wise DC wickets between 91st and 120th ball i.e. data between 16th and 20th over:

### Filtering CSK and Delhi matches where batting team is DC and bowling team is CSK:

In [20]:
dc_wickets = data2.loc[(data2['batting_team'] == "Delhi Capitals") & (data2['bowling_team'] == "Chennai Super Kings")]
dc_wickets.head()

match_id  inning    batting_team         bowling_team  over  ball  \
18498        79       2  Delhi Capitals  Chennai Super Kings     1     1   
18499        79       2  Delhi Capitals  Chennai Super Kings     1     2   
18500        79       2  Delhi Capitals  Chennai Super Kings     1     3   
18501        79       2  Delhi Capitals  Chennai Super Kings     1     4   
18502        79       2  Delhi Capitals  Chennai Super Kings     1     5   

         batsman non_striker   bowler  is_super_over  wide_runs  bye_runs  \
18498  G Gambhir    V Sehwag  M Ntini              0          0         0   
18499   V Sehwag   G Gambhir  M Ntini              0          0         0   
18500   V Sehwag   G Gambhir  M Ntini              0          0         0   
18501   V Sehwag   G Gambhir  M Ntini              0          0         0   
18502   V Sehwag   G Gambhir  M Ntini              0          0         0   

       legbye_runs  noball_runs  penalty_runs  batsman_runs  extra_runs  \
18498            0            0             0             1           0   
18499            0            0             0             0           0   
18500            0            0             0             0           0   
18501            0            0             0             6           0   
18502            0            0             0             0           0   

       total_runs player_dismissed dismissal_kind fielder  
18498           1              NaN            NaN     NaN  
18499           0              NaN            NaN     NaN  
18500           0              NaN            NaN     NaN  
18501           6              NaN            NaN     NaN  
18502           0              NaN            NaN     NaN

### Filtering data between 16th and 120th over:

In [21]:
dc_wickets = dc_wickets.loc[(dc_wickets['over'] >= 16) & (dc_wickets['over'] <= 20)]
dc_wickets = pd.DataFrame(dc_wickets.groupby('match_id')['player_dismissed'].count())
dc_wickets = dc_wickets.rename({'player_dismissed' : 'No. of Wickets'}, axis=1)
dc_wickets

No. of Wickets
match_id                
79                     0
87                     3
125                    1
141                    5
185                    2
224                    0
289                    3
364                    0
380                    3
404                    3
446                    3
465                    1
483                    0
519                    4
565                    1
7923                   1
7945                   0
11141                  4
11341                  2
11414                  4

### Average Wickets:

In [22]:
avg_wic = dc_wickets['No. of Wickets'].mean()
avg_wic

2.0

**Answer:** On an average, CSK takes 2 wickets between 91st and 120th ball.

## Question 4: How many balls will CSK take to score their 100 runs?

In [37]:
csk_balls = data2.loc[(data2['batting_team'] == "Chennai Super Kings")]
csk_balls.head(20)

match_id  inning         batting_team     bowling_team  over  ball  \
14087        61       1  Chennai Super Kings  Kings XI Punjab     1     1   
14088        61       1  Chennai Super Kings  Kings XI Punjab     1     2   
14089        61       1  Chennai Super Kings  Kings XI Punjab     1     3   
14090        61       1  Chennai Super Kings  Kings XI Punjab     1     4   
14091        61       1  Chennai Super Kings  Kings XI Punjab     1     5   
14092        61       1  Chennai Super Kings  Kings XI Punjab     1     6   
14093        61       1  Chennai Super Kings  Kings XI Punjab     2     1   
14094        61       1  Chennai Super Kings  Kings XI Punjab     2     2   
14095        61       1  Chennai Super Kings  Kings XI Punjab     2     3   
14096        61       1  Chennai Super Kings  Kings XI Punjab     2     4   
14097        61       1  Chennai Super Kings  Kings XI Punjab     2     5   
14098        61       1  Chennai Super Kings  Kings XI Punjab     2     6   
14099        61       1  Chennai Super Kings  Kings XI Punjab     2     7   
14100        61       1  Chennai Super Kings  Kings XI Punjab     3     1   
14101        61       1  Chennai Super Kings  Kings XI Punjab     3     2   
14102        61       1  Chennai Super Kings  Kings XI Punjab     3     3   
14103        61       1  Chennai Super Kings  Kings XI Punjab     3     4   
14104        61       1  Chennai Super Kings  Kings XI Punjab     3     5   
14105        61       1  Chennai Super Kings  Kings XI Punjab     3     6   
14106        61       1  Chennai Super Kings  Kings XI Punjab     3     7   

          batsman non_striker       bowler  is_super_over  wide_runs  \
14087    PA Patel   ML Hayden        B Lee              0          0   
14088    PA Patel   ML Hayden        B Lee              0          0   
14089    PA Patel   ML Hayden        B Lee              0          0   
14090   ML Hayden    PA Patel        B Lee              0          0   
14091   ML Hayden    PA Patel        B Lee              0          0   
14092   ML Hayden    PA Patel        B Lee              0          0   
14093    PA Patel   ML Hayden  S Sreesanth              0          0   
14094    PA Patel   ML Hayden  S Sreesanth              0          0   
14095    PA Patel   ML Hayden  S Sreesanth              0          0   
14096    PA Patel   ML Hayden  S Sreesanth              0          0   
14097    PA Patel   ML Hayden  S Sreesanth              0          1   
14098    PA Patel   ML Hayden  S Sreesanth              0          0   
14099    PA Patel   ML Hayden  S Sreesanth              0          0   
14100   ML Hayden    PA Patel        B Lee              0          0   
14101   ML Hayden    PA Patel        B Lee              0          0   
14102   ML Hayden    PA Patel        B Lee              0          0   
14103   ML Hayden    PA Patel        B Lee              0          0   
14104    PA Patel   ML Hayden        B Lee              0          0   
14105  MEK Hussey   ML Hayden        B Lee              0          0   
14106   ML Hayden  MEK Hussey        B Lee              0          0   

       bye_runs  legbye_runs  noball_runs  penalty_runs  batsman_runs  \
14087         0            0            0             0             0   
14088         0            0            0             0             0   
14089         0            0            0             0             1   
14090         0            0            0             0             0   
14091         0            0            0             0             4   
14092         0            0            0             0             0   
14093         0            0            0             0             4   
14094         0            0            0             0             4   
14095         0            0            0             0             4   
14096         0            0            0             0             2   
14097         0            0            0             0             0   


In [58]:
csk_balls.pivot_table(values='total_runs',
                     index = 'over',
                     columns = 'match_id',
                     aggfunc='sum')

match_id  61     67     71     74     79     83     85     87     91     \
over                                                                      
1           5.0    7.0    5.0    1.0    4.0    1.0    4.0    9.0   12.0   
2          15.0    9.0    3.0   10.0    7.0    9.0    6.0   15.0    8.0   
3           7.0   18.0   10.0    7.0    2.0    2.0    4.0    4.0    2.0   
4           6.0    1.0    7.0    4.0   15.0    4.0    7.0    6.0    5.0   
5           9.0    5.0   10.0    3.0    9.0   14.0   10.0    8.0   11.0   
6          11.0   13.0   10.0   11.0   10.0    8.0    3.0   20.0   15.0   
7           9.0    6.0   10.0    7.0    7.0    3.0    9.0    9.0   11.0   
8           8.0    5.0   11.0    7.0   11.0    5.0    7.0   13.0    4.0   
9          13.0    8.0    6.0   10.0    5.0    5.0   10.0   11.0   10.0   
10          5.0   10.0    5.0    5.0   12.0    7.0    6.0    3.0    7.0   
11         19.0   14.0    4.0    7.0   16.0    1.0    4.0    2.0   15.0   
12         20.0   16.0    9.0   11.0    8.0   17.0    7.0    7.0    9.0   
13          6.0   11.0    5.0    3.0    8.0    8.0    4.0    8.0   15.0   
14         10.0   14.0   12.0    8.0    5.0    3.0    3.0   23.0    8.0   
15         18.0    6.0   14.0    8.0    7.0    4.0    2.0    4.0   12.0   
16          8.0   15.0   18.0   14.0    1.0    5.0    9.0    6.0    5.0   
17         14.0   10.0   13.0   16.0   10.0    1.0   12.0    7.0    5.0   
18         17.0   14.0    NaN    8.0    7.0    1.0   14.0   11.0    5.0   
19         15.0   12.0    NaN   24.0   14.0   11.0    3.0    7.0   17.0   
20         25.0   14.0    NaN   14.0   11.0    NaN   20.0   15.0    5.0   

match_id  96     103    107    110    114    116    117    118    122    \
over                                                                      
1           1.0    5.0    5.0    6.0   12.0    7.0    6.0    2.0    9.0   
2          12.0   13.0    2.0    9.0    6.0    7.0    6.0   14.0    6.0   
3           5.0    9.0   20.0   19.0    4.0    2.0    4.0    2.0   13.0   
4           2.0    3.0   13.0    6.0   13.0    9.0   13.0    1.0    8.0   
5           3.0    6.0    2.0   15.0    6.0    5.0   10.0   14.0   20.0   
6           5.0    7.0    2.0   10.0    3.0    5.0    3.0    7.0   12.0   
7           8.0    4.0    9.0    7.0    9.0    9.0    9.0   15.0    9.0   
8           6.0    8.0    3.0    9.0    6.0    9.0    6.0    2.0    8.0   
9           4.0    NaN    4.0    3.0    9.0    7.0    7.0    9.0   12.0   
10         11.0    NaN    1.0   15.0    4.0    2.0   11.0    4.0    9.0   
11          0.0    NaN    4.0   13.0    5.0   13.0    2.0    3.0    2.0   
12          9.0    NaN    4.0   17.0   11.0    5.0   10.0   10.0    3.0   
13          7.0    NaN   15.0   12.0    8.0   10.0   10.0    8.0    5.0   
14         10.0    NaN    3.0    3.0   12.0   10.0   12.0    3.0   13.0   
15          6.0    NaN    5.0   13.0    6.0   16.0    8.0   14.0    6.0   
16         13.0    NaN    1.0    8.0    3.0    NaN   10.0    6.0    9.0   
17         11.0    NaN    5.0   12.0    3.0    NaN    3.0   13.0    7.0   
18         14.0    NaN   11.0    8.0   11.0    NaN    9.0    8.0    7.0   
19         14.0    NaN    2.0   12.0   10.0    NaN    9.0    6.0    7.0   
20         15.0    NaN    1.0    4.0    7.0    NaN   15.0    5.0   14.0   

match_id  125    131    137    141    144    149    152    159    162    \
over                                                                      
1          11.0    2.0    4.0   10.0   10.0    5.0    9.0    4.0    4.0   
2          10.0   16.0    1.0    6.0    5.0    8.0    1.0   16.0   13.0   
3           9.0   10.0   12.0    3.0    6.0   16.0    9.0   12.0    1.0   
4          14.0   10.0    2.0    9.0   13.0    6.0    4.0    0.0    6.0   
5          12.0   17.0    7.0    8.0   11.0   10.0    2.0   11.0   10.0   
6           5.0    7.0   14.0   17.0    8.0   12.0   12.0    9.0    1.0   
7          24.0    3.0    7.0    5.0    8.0    5.0    3.0    2.0    4.0   
8           5.0    4.0 

## Question 5: How many no-balls will be bowled in the match?

#### Solution: Filtering data with CSK and Delhi Matches:

In [66]:
csk_dc_nb = data2.loc[((data2['batting_team'] == "Chennai Super Kings") & (data2['bowling_team'] == "Delhi Capitals") | (data2['batting_team'] == "Delhi Capitals") & (data2['bowling_team'] == "Chennai Super Kings"))]
csk_dc_nb = pd.DataFrame(csk_dc_nb.groupby('match_id')['noball_runs'].sum())
csk_dc_nb

noball_runs
match_id             
79                  4
87                  2
125                 1
141                 1
185                 1
224                 2
289                 1
318                 0
364                 0
380                 0
404                 0
446                 0
465                 0
483                 0
519                 1
565                 0
7923                1
7945                0
11141               1
11341               1
11414               1

#### Getting mean of the no-ball runs match wise:

In [67]:
avg_nb = csk_dc_nb['noball_runs'].mean()
avg_nb

0.8095238095238095

**Answer:** Average no-ball run in each match is very less 0.81. So, I would prdict that around 1-2 no-balls will be bowled in this match.